hazelcast/src/main/java/com/hazelcast/spi/impl/SlowOperationLog.java

```diff
@@ -171,36 +173,37 @@ public void fromJson(JsonObject json) {
     }
 
     static final class Invocation implements JsonSerializable {
+        // values are just set once, but cannot be final due to JsonSerializable
         int id;
         long startedAt;
-        long durationNanos;
-        volatile long lastAccessNanos;
 
-        public long getDurationNanos() {
-            return durationNanos;
+        // only accessed by single SlowOperationDetectorThread, no need for synchronization
+        long lastAccessNanos;
+
+        // this field will be changed by single SlowOperationDetectorThread only (they can be read by multiple threads)
+        volatile int durationMs;
+
+        public long getDurationMs() {
+            return durationMs;
         }
 
         @Override
         public String toString() {
-            // NOP to read lastAccess to update local thread values
-            if (lastAccessNanos < 0) {
-                assert true;
-            }
+            // read durationMs first to enforce a read barrier
+            int tmpDurationMs = durationMs;
 
-            return "Invocation{" + "id=" + id + ", startedAt=" + startedAt + ", durationNanos=" + durationNanos + '}';
+            return "Invocation{" + "id=" + id + ", startedAt=" + startedAt + ", durationMs=" + tmpDurationMs + '}';
         }
 
         @Override
         public JsonObject toJson() {
-            // NOP to read lastAccess to update local thread values
-            if (lastAccessNanos < 0) {
-                assert true;
-            }
+            // read durationMs first to enforce a read barrier
+            int tmpDurationMs = durationMs;
```

Comment:
To me, visibility of effectively final fields of an object should be guaranteed via safe publication of object reference itself. There should be no need to rely on any read ordering in methods like `toString()` or `toJson()`.